Add london_underground

robmarkcole opened the pull request:
## Description:
This sensor displays the current status of London Underground tube lines. If there is a disruption, a description of the disruption is viewable as an attribute. 
This pull request replaces https://github.com/home-assistant/home-assistant/pull/8235

## Example entry for `configuration.yaml`:
```yaml
sensor:
  - platform: london_underground
    line:
      - Bakerloo
      - Central
      - Circle
      - District
      - DLR
      - Hammersmith & City
      - Jubilee
      - London Overground
      - Metropolitan
      - Northern
      - Piccadilly
      - TfL Rail
      - Victoria
      - Waterloo & City
```

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) (if applicable):** https://github.com/home-assistant/home-assistant.github.io/pull/2902

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

robmarkcole committed:
Add tube_state

Add tube_state sensor

balloob commented on homeassistant/components/sensor/tube_state.py:
```diff
@@ -0,0 +1,136 @@
+    'Waterloo & City']
+URL = 'https://api.tfl.gov.uk/line/mode/tube,overground,dlr,tflrail/status'
+
+CONFIG_SCHEMA = vol.Schema({
```
You are a platform so please define a platform schema:

```python
PLATFORM_SCHEMA = vol.Schema({
    vol.Required(CONF_LINE): vol.In(TUBE_LINES),
})
```

balloob commented on homeassistant/components/sensor/tube_state.py:
```diff
@@ -0,0 +1,136 @@
+
+ATTRIBUTION = "Powered by TfL Open Data"
+CONF_LINE = 'line'
+DOMAIN = 'tube_state'
```
platforms don't have a domain. Please remove this line.

balloob commented on homeassistant/components/sensor/tube_state.py:
```diff
@@ -0,0 +1,136 @@
+        sensors.append(LondonTubeSensor(line, data))
+
+    add_devices(sensors, True)
+    _LOGGER.info("The tube_state component is ready!")
```
Please remove these two log lines.

balloob commented on homeassistant/components/sensor/tube_state.py:
```diff
@@ -0,0 +1,136 @@
+    def update(self):
+        """Get the latest data from TFL."""
+        response = requests.get(URL)
+        _LOGGER.info("TFL Request made")
```
Please don't log this as info. If it needs to be logged at all, make it `debug`

robmarkcole committed:
Final cleanup

balloob commented on tests/components/sensor/test_tube_state.py:
```diff
@@ -0,0 +1,46 @@
+        """Stop everything that was started."""
+        self.hass.stop()
+
+    def test_setup_with_config(self):
```
Please merge this test with the one below. So instead of `tube_state.setup_platform(…)` just call `setup_component(…)`.

Then you can read the state from the state machine.

```python
state = self.hass.states.get('sensor.london_overground')
assert state.state == 'bla'
assert state.attributes.get('Description') == 'Something'
```

robmarkcole committed:
Make corrections

Correct PLATFORM_SCHEMA

robmarkcole committed:
Fix space

balloob commented:
🎉 🎉 🐬

robmarkcole committed:
Make test pass

robmarkcole commented:
There is an issue in test_tube_state.py that the following failed:

```
@requests_mock.Mocker()
def test_setup(self, mock_req):
    """Test for operational WSDOT sensor with proper attributes."""
    mock_req.get(URL, text=load_fixture('tube_state.json'))
    tube_state.setup_platform(self.hass, self.config, self.add_entities)
    self.assertEqual(len(self.entities), 1)
    self.assertTrue(
        setup_component(self.hass, 'sensor', {'tube_state': self.config}))
    state = self.hass.states.get('sensor.london_overground')
    assert state.state == 'Minor Delays'
```
with 

`AttributeError: 'NoneType' object has no attribute 'state'`

So I have:

```
    @requests_mock.Mocker()
    def test_setup(self, mock_req):
        """Test for operational WSDOT sensor with proper attributes."""
        mock_req.get(URL, text=load_fixture('tube_state.json'))
        tube_state.setup_platform(self.hass, self.config, self.add_entities)
        self.assertEqual(len(self.entities), 1)
        self.assertTrue(
            setup_component(self.hass, 'sensor', {'tube_state': self.config}))
        sensor = self.entities[0]
        self.assertEqual(sensor.name, 'London Overground')
        self.assertEqual(sensor.state, 'Minor Delays')
```

which is how darksky tests

robmarkcole commented:
Re travis, `The command "travis_wait tox" exited with 1.`

robmarkcole committed:
Correct format of test

Test still failing, don’t understand why

robmarkcole committed:
correct description

lwis commented:
@robmarkcole `self.hass.states.get('sensor.london_overground')` is returning None.

robmarkcole commented:
Aha I have  update_before_add=False

robmarkcole commented:
Test does pass with:

```
        sensor = self.entities[0]
        self.assertEqual(sensor.name, 'London Overground')
        self.assertEqual(sensor.state, 'Minor Delays')
```

robmarkcole committed:
Make test pass

Preferred method below returns None

state = self.hass.states.get('sensor.london_overground')

robmarkcole committed:
Format for hound

robmarkcole committed:
indent

robmarkcole commented on tests/components/sensor/test_tube_state.py:
```diff
@@ -0,0 +1,49 @@
+        self.assertEqual(len(self.entities), 1)
+        self.assertTrue(
+            setup_component(self.hass, 'sensor', {'tube_state': self.config}))
+        # state = self.hass.states.get('sensor.london_overground')
```
Returns None

balloob replied:
Instead of commenting it out, check what the entity name is. It's right there in the test logs.

robmarkcole commented on tests/components/sensor/test_tube_state.py:
```diff
@@ -0,0 +1,49 @@
+        # assert state.state == 'Minor Delays'
+        # assert state.attributes.get('Description') == 'something'
+
+        sensor = self.entities[0]
```
This passes. Copied technique from test_wsdot.py

balloob replied:
This technique should not be used. please remove.

balloob commented on tests/components/sensor/test_tube_state.py:
```diff
@@ -0,0 +1,49 @@
+class TestLondonTubeSensor(unittest.TestCase):
+    """Test the tube_state platform."""
+
+    def add_entities(self, new_entities, update_before_add=False):
```
Remove this.

balloob commented on tests/components/sensor/test_tube_state.py:
```diff
@@ -0,0 +1,49 @@
+        """Initialize values for this testcase class."""
+        self.hass = get_test_home_assistant()
+        self.config = {CONF_LINE: ['London Overground']}
+        self.entities = []
```
Remove this.

balloob commented on tests/components/sensor/test_tube_state.py:
```diff
@@ -0,0 +1,49 @@
+    def test_setup(self, mock_req):
+        """Test for operational tube_state sensor with proper attributes."""
+        mock_req.get(URL, text=load_fixture('tube_state.json'))
+        tube_state.setup_platform(self.hass, self.config, self.add_entities)
```
Remove this.

balloob commented on tests/components/sensor/test_tube_state.py:
```diff
@@ -0,0 +1,49 @@
+        """Test for operational tube_state sensor with proper attributes."""
+        mock_req.get(URL, text=load_fixture('tube_state.json'))
+        tube_state.setup_platform(self.hass, self.config, self.add_entities)
+        self.assertEqual(len(self.entities), 1)
```
And this.

balloob reviewed (changes requested):
The test needs fixing.

robmarkcole committed:
Make requested changes to test, not working

Test fails with:

AssertionError: assert 0 > 0
where 0 = len([])

Surely I need tube_state.setup_platform ?

robmarkcole commented on tests/components/sensor/test_tube_state.py:
```diff
@@ -0,0 +1,34 @@
+            setup_component(self.hass, 'sensor', {'tube_state': self.config}))
+
+        ids = self.hass.states.entity_ids()
+        assert len(ids) > 0
```
I have zero entities

robmarkcole replied:
Test log on https://github.com/robmarkcole/London-tube-status/blob/master/development/tox_output.txt

robmarkcole committed:
Fixed test

Config was wrong

robmarkcole committed:
Change component name to london_tube

robmarkcole committed:
Update name to london_underground

Make consistent

robmarkcole committed:
cleanup

robmarkcole commented:
I'm not sure what needs to be done about the failing Travis @ https://travis-ci.org/home-assistant/home-assistant/jobs/249272537

balloob commented:
Woohooo 🎉

robmarkcole commented:
Fantastic 🤗
